### PR TITLE
Add minimal eval support with redis.call

### LINKED ISSUE #63: support redis eval script (lua)
Can we do support the EVAL command?

https://redis.io/commands/eval

### PR DESCRIPTION
**Pull request by Negashev**
Fix #63 

Work with simple call
```bash
127.0.0.1:6379> eval "return redis.call('set','foo','bar')" 0
OK
127.0.0.1:6379> GET foo
"bar"
127.0.0.1:6379> eval "return redis.call('GET','foo')" 0
bar
127.0.0.1:6379> eval "return 10" 0
(integer) 10
127.0.0.1:6379> eval "return {1,2,3.3333,'foo',nil,'bar'}" 0 <--- some problems (redis return only 4 keys)
1) (integer) 1
2) (integer) 2
3) (integer) 3
4) foo
5) bar
127.0.0.1:6379> eval "return {1,2,{3,'Hello World!'}}" 0
1) (integer) 1
2) (integer) 2
3) 1) (integer) 3
   2) Hello World!
127.0.0.1:6379> eval "return redis.call(KEYS[1],KEYS[2], 134)" 2 set foo
OK
127.0.0.1:6379> GET foo
"134"
127.0.0.1:6379> eval "return {KEYS[1],KEYS[2],ARGV[1],ARGV[2]}" 2 key1 key2 first second
1) key1
2) key2
3) first
4) second
```

Ready

- [x] `redis.call` from the box
- [x] `KEYS` and `ARGV` in lua
- [x] connect to titan internal
- [ ] make transactional (help wanted)

### CODE REVIEW

**Comment by arthurkiller** [comment]
Really cool! Thx for your contribution.

Have an awesome day

✉️✉️✉️✉️✉️✉️✉️✉️✉️✉️✉️✉️
Arthur lee

File: command/scripting.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+	"github.com/meitu/titan/encoding/resp"
+	"github.com/yuin/gopher-lua"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var cli *context.ClientContext
+
+func NewClientContext(ctx *Context) *context.ClientContext {
```
[the remark this answers is not in the document]
Comment (Negashev) [reply command/scripting.go]: @YIDWang Do not quite understand what you mean (I am not strong in golang)

File: command/scripting.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	if errInteger == nil {
+		L.Push(lua.LNumber(valInteger))
+		return 1
+	}
+
+	return 0
+}
+
+// Eval lua script the server
+func Eval(ctx *Context) {
```
Comment (shafreeck) [inline command/scripting.go]: Eval can be used in multi-exec block in Redis, so it is a TxnCommand

File: command/scripting.go [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+		resp.ReplyError(ctx.Out, ErrWrongArgs(ctx.Name).Error())
+		return
+	}
+
+	luaScript := fmt.Sprintf(args[0][:])
+	// replace KEYS for lua
+	keysArray := make([]string, len(args[2:keysLen]))
+	if keysLen > 2 {
+		for key := range keysArray {
+			luaScript = strings.Replace(luaScript, fmt.Sprintf("KEYS[%d]", key+1), fmt.Sprintf("'%s'", string(args[key+2])), 1)
```
Comment (shafreeck) [inline command/scripting.go]: Why replace `KEYS`, `ARGS` here? Should they be global variables?

File: command/scripting.go [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+	} else {
+		resp.ReplyNullBulkString(ctx.Out)
+	}
+}
+
+// Constructor
+func newRedisCall(L *lua.LState) int {
+	var rest []string
+	// filter data from lua to redis command without first arg
+	GetLuaTop := L.GetTop()
```
Comment (shafreeck) [inline command/scripting.go]: What if GetLuaTop < 2 ?
Return an error when there is no arguments according to redis:
>ERR Error running script (call to f_0a907e1429221a4d85516cab7fd219a82a9439d8): @user_script:1: @user_script: 1: Please specify at least one argument for redis.call()

File: command/scripting.go [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+	out := bytes.NewBuffer(nil)
+	ctx := &Context{
+		Name:    L.Get(1).String(),
+		Args:    rest,
+		In:      nil,
+		Out:     out,
+		Context: context.New(cli, serv),
+	}
+
+	Call(ctx)
```
Comment (shafreeck) [inline command/scripting.go]: There may be multiple`redis.call` in a Lua script, so using `TxnCall` instead

File: command/scripting.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+	data := out.String()
+
+	//return array of string
+	dArray := resp.NewDecoder(bytes.NewBufferString(data))
```
Comment (shafreeck) [inline command/scripting.go]: It is a little wired to parse the returned value lying on error handling, maybe we should check each type of value here.

**Comment by arthurkiller** [comment]
@Negashev  hi dude, still here?

We are planning to merge this pr into another branch and resolve the comments.

**Comment by Negashev** [comment]
@arthurkiller Hi, I can not say with certainty that this is govot. I still had a lot of questions during the implementation, as I lack experience with golang, so I stopped at https://github.com/JohnSully/KeyDB

**Comment by arthurkiller** [comment]
@Negashev  Sorry for hearing that, why did you abandoned on titan?

I will close this pr then, and the `script` will be implemented into titan soon.
thx for your contribute anyway.

**Comment by Negashev** [comment]
Titan is a wonderful product, in the issue of redis distribution, but still it is convenient only for new projects.

**Pull request closed without merging**